core/state/statedb_utils.go

```diff
@@ -0,0 +1,145 @@
+package state
+
+import (
+	"math/big"
+
+	"github.com/XinFinOrg/XDPoSChain/common"
+	"github.com/XinFinOrg/XDPoSChain/core/types"
+
+	"github.com/XinFinOrg/XDPoSChain/crypto"
+)
+
+var (
+	slotBlockSignerMapping = map[string]uint64{
+		"blockSigners": 0,
+		"blocks":       1,
+	}
+)
+
+func GetSigners(statedb *StateDB, block *types.Block) []common.Address {
+	slot := slotBlockSignerMapping["blockSigners"]
+	keys := []common.Hash{}
+	keyArrSlot := GetLocMappingAtKey(block.Hash(), slot)
+	arrSlot := statedb.GetState(common.HexToAddress(common.BlockSigners), common.BigToHash(keyArrSlot))
+	arrLength := arrSlot.Big().Uint64()
+	for i := uint64(0); i < arrLength; i++ {
+		key := GetLocDynamicArrAtElement(common.BigToHash(keyArrSlot), i, 1)
+		keys = append(keys, key)
+	}
+	rets := []common.Address{}
+	for _, key := range keys {
+		ret := statedb.GetState(common.HexToAddress(common.BlockSigners), key)
+		rets = append(rets, common.HexToAddress(ret.Hex()))
+	}
+
+	return rets
+}
+
+var (
+	slotRandomizeMapping = map[string]uint64{
+		"randomSecret":  0,
+		"randomOpening": 1,
+	}
+)
+
+func GetSecret(statedb *StateDB, address common.Address) [][32]byte {
+	slot := slotRandomizeMapping["randomSecret"]
+	locSecret := GetLocMappingAtKey(address.Hash(), slot)
+	arrLength := statedb.GetState(common.HexToAddress(common.RandomizeSMC), common.BigToHash(locSecret))
+	keys := []common.Hash{}
+	for i := uint64(0); i < arrLength.Big().Uint64(); i++ {
+		key := GetLocDynamicArrAtElement(common.BigToHash(locSecret), i, 1)
+		keys = append(keys, key)
+	}
+	rets := [][32]byte{}
+	for _, key := range keys {
+		ret := statedb.GetState(common.HexToAddress(common.RandomizeSMC), key)
+		rets = append(rets, ret)
+	}
+	return rets
+}
+
+func GetOpening(statedb *StateDB, address common.Address) [32]byte {
+	slot := slotRandomizeMapping["randomOpening"]
+	locOpening := GetLocMappingAtKey(address.Hash(), slot)
+	ret := statedb.GetState(common.HexToAddress(common.RandomizeSMC), common.BigToHash(locOpening))
+	return ret
+}
+
+var (
+	slotValidatorMapping = map[string]uint64{
+		"withdrawsState":         0,
+		"validatorsState":        1,
+		"voters":                 2,
+		"candidates":             3,
```

Comment:
I tested the following state db util functions and some of them didn’t work. It’s the `slotValidatorMapping` not compatible with the current smart contract. I see that in this PR, smart contract is changed, but I’m afraid that the existing storage is still incompatible with these state db util functions.

Reply:
This is important for consensus since it is used when processing/validating checkpoint blocks.

Reply:
ValidatorMapping has been updated, added missing KYC mapping.